// ngx-typeahead/esm5/lib/ngx-typeahead.module.js
import { __decorate } from "tslib";
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { HttpClientModule, HttpClientJsonpModule } from '@angular/common/http';
import { NgxTypeAheadComponent } from './ngx-typeahead.component';
import * as ɵngcc0 from '@angular/core';
var NgxTypeaheadModule = /** @class */ (function () {
    function NgxTypeaheadModule() {
    }
NgxTypeaheadModule.ɵfac = function NgxTypeaheadModule_Factory(t) { return new (t || NgxTypeaheadModule)(); };
NgxTypeaheadModule.ɵmod = /*@__PURE__*/ ɵngcc0.ɵɵdefineNgModule({ type: NgxTypeaheadModule });
NgxTypeaheadModule.ɵinj = /*@__PURE__*/ ɵngcc0.ɵɵdefineInjector({ imports: [[CommonModule, HttpClientModule, HttpClientJsonpModule], CommonModule] });
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && ɵngcc0.ɵsetClassMetadata(NgxTypeaheadModule, [{
        type: NgModule,
        args: [{
                declarations: [NgxTypeAheadComponent],
                imports: [CommonModule, HttpClientModule, HttpClientJsonpModule],
                exports: [NgxTypeAheadComponent, CommonModule]
            }]
    }], function () { return []; }, null); })();
(function () { (typeof ngJitMode === "undefined" || ngJitMode) && ɵngcc0.ɵɵsetNgModuleScope(NgxTypeaheadModule, { declarations: function () { return [NgxTypeAheadComponent]; }, imports: function () { return [CommonModule, HttpClientModule, HttpClientJsonpModule]; }, exports: function () { return [NgxTypeAheadComponent, CommonModule]; } }); })();
    return NgxTypeaheadModule;
}());
export { NgxTypeaheadModule };
